packages/java/endpoint/src/main/java/dev/hilla/EndpointCodeGenerator.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2000-2022 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package dev.hilla;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+import dev.hilla.engine.EngineConfiguration;
+import dev.hilla.engine.GeneratorProcessor;
+import dev.hilla.engine.ParserProcessor;
+import org.springframework.stereotype.Component;
+
+import com.vaadin.flow.server.VaadinContext;
+import com.vaadin.flow.server.frontend.FrontendTools;
+import com.vaadin.flow.server.startup.ApplicationConfiguration;
+
+/**
+ * Handles (re)generation of the TypeScript code.
+ */
+@Component
+public class EndpointCodeGenerator {
+
+    private final Path buildDirectory;
+    private final EndpointController endpointController;
+    private final ApplicationConfiguration configuration;
+    private final String nodeExecutable;
+
+    private static EndpointCodeGenerator instance;
+
+    /**
+     * Creates the singleton.
+     * 
+     * @param context
+     *            the context the application is running in
+     * @param endpointController
+     *            a reference to the endpoint controller
+     */
+    public EndpointCodeGenerator(VaadinContext context,
+            EndpointController endpointController) {
+        this.endpointController = endpointController;
+        configuration = ApplicationConfiguration.get(context);
+
+        Path projectFolder = configuration.getProjectFolder().toPath();
+        buildDirectory = projectFolder.resolve(configuration.getBuildFolder());
+
+        FrontendTools tools = new FrontendTools(configuration,
+                configuration.getProjectFolder());
+        this.nodeExecutable = tools.getNodeBinary();
+        instance = this;
+    }
+
+    /**
+     * Gets the singleton instance.
+     * 
+     * @throws IOException
+     */
+    public static EndpointCodeGenerator getInstance() {
+        return instance;
+    }
```

Comment:
I suggest avoiding the singleton anti-pattern. We should be able to manage passing a singleton instance through the Spring configuration and constructor injection where needed.

Reply:
The consumer of this would be e.g. a Hotswap agent plugin

Reply:
Well, nothing else can be done I guess.

Reply:
Hotswap agent cannot be aware of Spring Boot so it is better to do the hacky Spring context "lookup" inside here instead of in a HA plugin

Reply:
I removed it for now, as it is not strictly needed for this change